cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -343,35 +352,51 @@ private synchronized void writeData(long start) {
-      // future is completed successfully.
-      // If the batch publish fails, none of the messages are persisted,
-      // and the future is completed exceptionally, allowing the caller to retry the entire StoreRequest.
-      RequestFuture headRequest = batch.poll();
-      List<Mutation> mutationsForRequest = createMutations(headRequest.getStoreRequest(),
-          sequenceId);
-      mutations.addAll(mutationsForRequest);
-      futures.add(headRequest.getFuture());
-      estimatedBatchSize += headRequest.getSize();
-
-      // For better performance, wait for more messages to accumulate before committing the batch.
-      if (batch.isEmpty() && (sequenceId.get() < publishBatchSize
-          || System.currentTimeMillis() - start < publishBatchTimeoutMillis)) {
-        try {
-          Thread.sleep(publishDelayMillis);
-        } catch (InterruptedException e) {
-          for (CompletableFuture<Void> future : futures) {
-            future.completeExceptionally(e);
+    try {
+      // Check if adding mutations related to the head will exceed the batch size limit.
+      // If exceeding limit, publish the accumulated mutations and then only pick the next item in queue.
+      while (!batch.isEmpty() && (estimatedBatchSize + batch.peek().getSize()
+          < MAX_BATCH_SIZE_IN_BYTES)) {
+        // Ensure that all messages within a StoreRequest are published in a single batch to maintain atomicity.
+        // If the batch publish succeeds, all messages are persisted, and the corresponding
+        // future is completed successfully.
+        // If the batch publish fails, none of the messages are persisted,
+        // and the future is completed exceptionally, allowing the caller to retry the entire StoreRequest.
+        RequestFuture headRequest = batch.poll();
+        List<Mutation> mutationsForRequest = createMutations(headRequest.getStoreRequest(),
+            sequenceId);
+        mutations.addAll(mutationsForRequest);
+        futures.add(headRequest.getFuture());
+        estimatedBatchSize += headRequest.getSize();
+
+        // For better performance, wait for more messages to accumulate before committing the batch.
+        if (batch.isEmpty() && (sequenceId.get() < publishBatchSize
+            || System.currentTimeMillis() - start < publishBatchTimeoutMillis)) {
+          try {
+            Thread.sleep(publishDelayMillis);
+          } catch (InterruptedException e) {
+            completeExceptionally(mutations, futures, e);
+            return;
           }
         }
       }
+      commitPublishMutations(mutations, futures);
+      futures.clear();
+    } finally {
+      // If any other exception happens (RuntimeException / Error, e.g. StackOverflow) apart from the
+      // handled InterruptedException and SpannerException, all waiting threads should not be stuck forever.
+      completeExceptionally(mutations, futures, new IOException("Publish operation failed"));
```

Comment:
completeExceptionally in finally block looks very odd. Rename to completeBlockedFutures.